radix-engine-stores/src/hash_tree/mod.rs

```diff
@@ -191,16 +210,25 @@ fn put_changes<S: TreeStore<P>, P: Payload>(
     root_hash
 }
 
-fn to_partition_change(
-    change: IdChange<DbPartitionKey, TreeRoot<()>>,
-) -> LeafChange<PartitionPayload> {
+fn to_node_change(change: IdChange<DbNodeKey, NodePayload>) -> LeafChange<NodePayload> {
     LeafChange {
-        key: LeafKey { bytes: change.id.0 },
-        new_payload: change.changed.map(|root| (root.hash, root.node)),
+        key: LeafKey::new(&change.id),
+        new_payload: change
+            .changed
+            .map(|payload| (calculate_node_hash(&payload), payload)),
     }
 }
 
-fn to_substate_change(change: IdChange<DbSortKey, Hash>) -> LeafChange<()> {
+fn calculate_node_hash(node_payload: &NodePayload) -> Hash {
```

Comment:
this is an interesting place:

When any of the Substates of a Node changes, its Partition root hash changes (it happens via regular JMT update), and then its Node root hash also must change (I made it happen myself, and here is the logic that calculates the new hash).

The calculation you see here only does 1x `hash()`, but I now wonder if it rather should do some funny JMT-like (or at least Merkle-like?) hierarchical hashing, in order to allow for smaller proofs...
Any thoughts welcome.

Reply:
Yes, I feel like this is potentially an overhead. Would it make sense to just make this middle tier a JMT?

Reply:
In terms of "compute merkle hash" (and "generate proof") operation, the 2-level JMT is same as 8-level Merkle tree (i.e. outputs 8 uncle-hashes). Does the average ReNode have >8 partitions?

Anyways: storing this embedded `Map<PartitionNumber, SubstateRootHash>` encoded as a set of JMT nodes would really be a pity (because there is a more compact and easier to understand encoding - the `Map<>`).

However, we can totally interpret the `Map` as a Merkle tree (a.k.a. "do some funny JMT-like (or at least Merkle-like?) hierarchical hashing`). Maybe I can even use the JMT's original methods (i.e. [this](https://github.com/radixdlt/radixdlt-scrypto/blob/99a674e5151565a7bb1e3cacefc510a405e52c2b/radix-engine-stores/src/hash_tree/types.rs#L821-L825))...

Reply:
For A1 - bear in mind every user transaction is updating a node with like 160 partitions (the TransactionTracker) - so we need to optimise for this case too.

Personally, I'd opt for a permanent JMT for Q1:
=> The root node of the partition tree can be stored in the leaf of the node tree.
=> We can re-use code / implementation for our substate tree

Reply:
Indeed a permanent true JMT has the "consistency" advantages... I guess the only downsides I see are:
- _(performance, minor)_ guaranteed +2 roundtrips to DB (per partition touched in the commit)
- _(conceptual)_ using a JMT for a map of bounded [and TBH absolutely tiny] cardinality seems like an overkill

Let me give it a try - if it is simple/elegant enough from implementation PoV, then I can live with it 👍

Reply:
Well most updates will be in the user partition, which for scrypto components will be the only partition in that nibble, so it will only be +1 round trip to the DB. But for native components, you're right, this will be +2 round trips because most of these use multiple partitions under the same first nibble. 

We are also decreasing the average number of distinct partitions in the upper tree by about 3x, which is logarithmically -1/2 of a layer in the top tree :D.

But yes, it will have some impact... I wonder what size of map is required before this is cheaper - I thought it was surprisingly expensive? But maybe I'm misremembering or we were capturing the other slowdowns?

Reply:
> I wonder what size of map is required before this is cheaper - I thought it was surprisingly expensive? But maybe I'm misremembering or we were capturing the other slowdowns?

I remember our impact attribution was wrong all the way. I have 50% of the "true JMT" impl ready, so we'll be soon able to run some true comparison benchmarks.

Reply:
Finally done ✅